### PR TITLE
Enhanced logging: Save log to file, log start of CommandsChanged

### PR DESCRIPTION
**Pull request by gerhardol**
part of #4213 and #4753

Changes proposed in this pull request:
 - Save command log to file
 - Add "id" to commands, log start and stop
 
Screenshots before and after (if PR changes UI):
image

What did I do to test the code and ensure quality:
 - Manual testing
 
Has been tested on (remove any that don't apply):
 - Windows 10

### CODE REVIEW

File: GitCommands/Logging/CommandLogEntry.cs [hunk 1]
```diff
@@ -2,23 +2,62 @@
...
+            Command = command;
+            ExecutionStartTimestamp = executionStartTimestamp;
+            ExecutionEndTimestamp = executionEndTimestamp;
+            IsStart = isStart;
+
+            // Start of commands will only be logged if enhanced logging is set,
+            // so cmdNo (with reference to start) will only be not null for ends with enhanced logging
+            _enhanced = isStart || startCmd != null;
+
+            if (isStart)
```
Comment (gerhardol) [inline GitCommands/Logging/CommandLogEntry.cs]: CommandLogger locks here, so a separate lock is not added

File: GitUI/CommandsDialogs/BrowseDialog/FormGitLog.cs [hunk 2]
```diff
@@ -119,6 +123,30 @@ private void alwaysOnTopCheckBox_CheckedChanged(object sender, EventArgs e)
...
+                Title = Name,
+                DefaultExt = ".txt",
+                AddExtension = true
+            })
+            {
+                fileDialog.Filter =
+                    "All files (*.*)|*.*";
+                if (fileDialog.ShowDialog(this) == DialogResult.OK)
+                {
+                    File.WriteAllText(fileDialog.FileName, AppSettings.GitLog.ToString());
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/BrowseDialog/FormGitLog.cs]: Log to file could request an enhanced format, for instance tab separated output, time in separate column

**Comment by gerhardol** [comment]
The changes could be separated, but as other issues pop up, I commit what I had done already

**Comment by drewnoakes** [comment]
### Bonus points

I'd love to see which class of thread requested the execution ("UI" vs "Thread Pool"). This would help us understand where we're blocking the UI while waiting for git processes to complete.

We should split out `ExecutablePath` and `Arguments` properties so we can hide paths in the UI, potentially making them available via tooltips. For me it's generally enough to see `git.exe`. Callers also don't need to worry about quoting paths.

Extra information via `LogProcessEnd` arguments: 

- Process exit code (hinted at above)
- Number of input bytes provided
- Number of output bytes read (when available)
- Number of error bytes read (when available)

We could highlight non-zero exits in the UI. We would also highlight processes requested on the UI thread.

Another idea is to modify the `ToString` method so that the duration and other common values appear in a columns, making them easier to scan quickly. Something like:

```c#
$"{ExecutionStartTimestamp:O}  {Duration.Elapsed.TotalMilliseconds:0.0,-6} ms  {ExitCode}  {ThreadClass}  {Path.GetFileName(ExecutablePath)} {Arguments}"
```

I am getting close to a PR that has most (all?) process execution going through a single gateway, which is therefore a single place for writing such events to the log. This should compose nicely with your changes and any of the above. 

With the changes in my pending work it should also be possible to capture "time until first byte". I'm not sure if this'll be useful in practice.

If you don't want to tackle any of the stuff I've described here then don't worry, I will give it a shot in time. I can move these thoughts to another issue, such as #4753.

File: GitCommands/Settings/AppSettings.cs [hunk 3]
```diff
@@ -1466,6 +1466,12 @@ public static BranchOrdering BranchOrderingCriteria
             set => SetEnum("BranchOrderingCriteria", value);
         }
 
+        public static bool EnhancedGitLog
+        {
+            get => GetBool("EnhancedGitLog", false);
+            set => SetBool("EnhancedGitLog", value);
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitCommands/Settings/AppSettings.cs]: Separate setting as most users will not be interested in logging start of commands. The setting is only in this form anyway.
While I like being able to customize, I try to not add too many new settings.

**Comment by gerhardol** [comment]
> Another idea is to modify the ToString method so that the duration and other common values appear in a columns

Yes, at least for saving to file (for instance csv).

> If you don't want to tackle any of the stuff I've described here then don't worry, I will give it a shot in time.

I have been running something like this for some time, when I saw your comment some days ago and RussKie's issue, I thought it was better to submit a PR and have this as starting point. We can close this PR now and discuss in the issue or discuss here in the PR.
Anyway, I do not propose merging this now and will wait a little before doing something more. Who that continues is a later matter.

File: GitCommands/Settings/AppSettings.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RussKie) [reply GitCommands/Settings/AppSettings.cs]: I think it is fine to overhaul the UX here and support a single more. The UI should be simple but sufficiently rich to express various pieces of information. 
We can control the display of certain elements via toolstrip/ribbon/context menus.

**Comment by RussKie** [comment]
@gerhardol touching base to see if any more work is planned for this one.

**Comment by gerhardol** [comment]
@drewnoakes planned to do something better, so I put this as wip.
I have not planned anything more right now.

**Comment by gerhardol** [comment]
@drewnoakes 
What is the state of your enhancements? (No obligation, you have done great additions but I still do not want to redo something (likely worse) than what you have done anyway).

If "almost done", I will not do anything (maybe extract the save log part).
If "some time later" I will probably change according to your proposal (not Bonus). I want to have this in 3.0.

If option to save start should be removed, the command log buffer will be doubled.

**Comment by drewnoakes** [comment]
@gerhardol I do not want to block you on this so if you have the time and inclination, go for it. If not, I will get to it eventually :)

I took a break from the git process stuff to clear my head from a weird bug, but will get back to it after this revision grid work. That branch will likely conflict with whatever you do here, but I will work out how to merge it if you go ahead. My branch will have all git process execution going through a single place in the code, where these metrics may be easily reported.

Would also like to have this stuff in 3.0. Do we have a timeline for when we want to stop making large changes and start focussing on stability?

**Comment by RussKie** [comment]
I would like to have a RC by the end of the month.
I totally understand it is an ambitious timeline.

File: GitCommands/Settings/AppSettings.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitCommands/Settings/AppSettings.cs]: Removed the Enhanced mode in the latest update (doubled the buffer size)

**Comment by gerhardol** [comment]
Updated (see commit comment for a description)

Open:
I changed to only log once, so just incomplete commands are set to "start". No duplications, easier to see where commands are not logged.
However, the commands are logged as they were called, not when they were completed as before. The current Queue must be replaced with another structure that can reorder entries to  get the entries in completion order (if not done, the command ID should be removed).

**Comment by gerhardol** [comment]
> I would like to have a RC by the end of the month. I totally understand it is an ambitious timeline.

We need somewhere to discuss such matters...
But in the current state, there is not much to display for new users compared to 2.51 - changes under the hood is not very interesting for users.

**Comment by gerhardol** [comment]
The previous logging order was for command completion, the updated is command starting. It is harder to see the the order of processing the command responses, but in general it should be better. 

I will remove the command id, as it has minor purpose when start is not logged individually.
After that a squash merge.

**Comment by RussKie** [comment]
Absolutely, would love to have tighter collaboration.
I think I saw something in GH that could be used as a team collaboration
area, but haven't explored it. Or a gitter private channel, if such thing
exists...

**Comment by vbjay** [comment]
#4508

**Comment by vbjay** [comment]
https://help.github.com/articles/about-project-boards/

**Comment by gerhardol** [comment]
A few other changes in the last commit

**Pull request merged**